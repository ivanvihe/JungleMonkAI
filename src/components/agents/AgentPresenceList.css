.agent-presence-list {
  display: flex;
  flex-direction: column;
  gap: 12px;
}

.agent-presence-item {
  display: flex;
  align-items: center;
  gap: 14px;
  padding: 14px 16px;
  border-radius: 14px;
  background: rgba(10, 10, 10, 0.75);
  border: 1px solid rgba(255, 255, 255, 0.08);
  transition: border 0.2s ease, box-shadow 0.2s ease;
}

.agent-presence-item.is-online {
  border-color: rgba(102, 255, 102, 0.6);
  box-shadow: 0 10px 24px rgba(102, 255, 102, 0.12);
}

.agent-presence-item.is-error {
  border-color: rgba(255, 102, 102, 0.45);
  box-shadow: 0 10px 24px rgba(255, 102, 102, 0.1);
}

.agent-presence-item.is-offline {
  border-color: rgba(255, 255, 255, 0.06);
}

.agent-presence-item.is-loading {
  border-style: dashed;
  border-color: rgba(255, 255, 255, 0.14);
}

.agent-presence-avatar {
  width: 42px;
  height: 42px;
  display: grid;
  place-items: center;
  font-weight: 600;
  font-size: 16px;
  border-radius: 12px;
  color: #fff;
}

.agent-presence-info {
  flex: 1;
  display: flex;
  flex-direction: column;
  gap: 6px;
}

.agent-presence-role {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
  gap: 8px;
  margin-top: 4px;
}

.agent-presence-role label {
  display: flex;
  flex-direction: column;
  font-size: 11px;
  text-transform: uppercase;
  letter-spacing: 0.6px;
  gap: 4px;
  color: rgba(255, 255, 255, 0.65);
}

.agent-presence-role select,
.agent-presence-role input {
  background: rgba(18, 18, 18, 0.9);
  border: 1px solid rgba(255, 255, 255, 0.12);
  border-radius: 8px;
  color: #fff;
  padding: 6px 8px;
  font-size: 12px;
}

.agent-presence-role input::placeholder {
  color: rgba(255, 255, 255, 0.45);
}

.agent-presence-name {
  font-size: 14px;
  font-weight: 600;
}

.agent-presence-meta {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  gap: 10px;
  font-size: 12px;
  color: rgba(255, 255, 255, 0.7);
}

.agent-presence-provider {
  letter-spacing: 0.6px;
  text-transform: uppercase;
}

.agent-presence-latency {
  font-variant-numeric: tabular-nums;
}

.agent-presence-status {
  display: inline-flex;
  align-items: center;
  gap: 6px;
  padding: 4px 8px;
  border-radius: 999px;
  background: rgba(255, 255, 255, 0.08);
}

.agent-presence-status.status-online {
  background: rgba(102, 255, 102, 0.15);
  color: #a9ffba;
}

.agent-presence-status.status-error {
  background: rgba(255, 102, 102, 0.18);
  color: #ffb4b4;
}

.agent-presence-status.status-offline {
  color: rgba(255, 255, 255, 0.65);
}

.agent-presence-status.status-loading {
  background: rgba(255, 255, 255, 0.1);
}

.agent-presence-message {
  font-size: 12px;
  color: rgba(255, 255, 255, 0.55);
}

.agent-presence-actions {
  display: flex;
  flex-direction: column;
  gap: 6px;
}

.presence-action {
  min-width: 98px;
  border: 1px solid rgba(255, 255, 255, 0.16);
  background: rgba(20, 20, 20, 0.9);
  color: #fff;
  border-radius: 8px;
  padding: 6px 10px;
  font-size: 12px;
  cursor: pointer;
  transition: background 0.2s ease, border 0.2s ease;
}

.presence-action.is-active {
  background: rgba(102, 255, 102, 0.12);
  border-color: rgba(102, 255, 102, 0.5);
}

.presence-action.is-ghost {
  background: rgba(255, 255, 255, 0.06);
  border-color: rgba(255, 255, 255, 0.08);
}

.presence-action:hover {
  background: rgba(255, 255, 255, 0.1);
}

.presence-led {
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background: currentColor;
  box-shadow: 0 0 12px currentColor;
}

@media (max-width: 1280px) {
  .agent-presence-item {
    flex-direction: column;
    align-items: flex-start;
  }

  .agent-presence-actions {
    width: 100%;
    flex-direction: row;
    justify-content: flex-end;
  }
}
