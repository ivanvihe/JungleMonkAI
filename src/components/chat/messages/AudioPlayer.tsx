import React from 'react';
import { ChatTranscription } from '../../../core/messages/messageTypes';

interface AudioPlayerProps {
  src: string;
  title?: string;
  mimeType?: string;
  transcriptions?: ChatTranscription[];
}

export const AudioPlayer: React.FC<AudioPlayerProps> = ({ src, title, mimeType, transcriptions }) => (
  <div className="message-audio-player">
    <audio src={src} controls preload="metadata" title={title ?? 'Mensaje de audio'} />
    {transcriptions?.length ? (
      <details className="audio-transcriptions">
        <summary>Transcripciones ({transcriptions.length})</summary>
        <ul>
          {transcriptions.map(transcription => (
            <li key={transcription.id}>
              <span className="transcription-language">{transcription.language ?? 'es'}</span>
              <span className="transcription-text">{transcription.text}</span>
            </li>
          ))}
        </ul>
      </details>
    ) : null}
  </div>
);

export default AudioPlayer;
