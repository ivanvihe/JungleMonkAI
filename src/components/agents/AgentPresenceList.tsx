import React from 'react';
import { AgentDefinition } from '../../core/agents/agentRegistry';
import { AgentPresenceEntry, AgentPresenceStatus } from '../../core/agents/presence';
import './AgentPresenceList.css';

interface AgentPresenceListProps {
  agents: AgentDefinition[];
  presence: Map<string, AgentPresenceEntry>;
  onToggleAgent: (agentId: string) => void;
  onOpenConsole?: (agentId: string) => void;
  onRefreshAgent?: (agentId: string) => void | Promise<void>;
}

const STATUS_LABELS: Record<AgentPresenceStatus, string> = {
  online: 'Operativo',
  offline: 'En espera',
  error: 'Con incidencias',
  loading: 'Verificando…',
};

const getStatusClass = (status: AgentPresenceStatus): string => {
  switch (status) {
    case 'online':
      return 'is-online';
    case 'error':
      return 'is-error';
    case 'offline':
      return 'is-offline';
    default:
      return 'is-loading';
  }
};

const buildAvatarLabel = (agent: AgentDefinition): string => {
  const [first] = agent.name.trim();
  if (first) {
    return first.toUpperCase();
  }
  return agent.provider.slice(0, 1).toUpperCase();
};

export const AgentPresenceList: React.FC<AgentPresenceListProps> = ({
  agents,
  presence,
  onToggleAgent,
  onOpenConsole,
  onRefreshAgent,
}) => (
  <ul className="agent-presence-list">
    {agents.map(agent => {
      const entry = presence.get(agent.id) ?? { status: 'loading', lastChecked: null };
      const statusClass = getStatusClass(entry.status);
      const latencyInfo =
        entry.latencyMs !== undefined ? `${entry.latencyMs} ms` : agent.kind === 'local' ? '∼ local' : '—';

      return (
        <li key={agent.id} className={`agent-presence-item ${statusClass}`}>
          <div
            className="agent-presence-avatar"
            style={{ background: `linear-gradient(135deg, ${agent.accent}, rgba(255, 255, 255, 0.08))` }}
            aria-hidden
          >
            {buildAvatarLabel(agent)}
          </div>
          <div className="agent-presence-info">
            <div className="agent-presence-name">{agent.name}</div>
            <div className="agent-presence-meta">
              <span className="agent-presence-provider">{agent.provider}</span>
              <span className="agent-presence-latency">{latencyInfo}</span>
              <span className={`agent-presence-status status-${entry.status}`}>
                <span className="presence-led" aria-hidden />
                {STATUS_LABELS[entry.status]}
              </span>
            </div>
            {entry.message && <div className="agent-presence-message">{entry.message}</div>}
          </div>
          <div className="agent-presence-actions">
            <button
              type="button"
              className={`presence-action ${agent.active ? 'is-active' : ''}`}
              onClick={() => onToggleAgent(agent.id)}
            >
              {agent.active ? 'Desactivar' : 'Activar'}
            </button>
            <button
              type="button"
              className="presence-action"
              onClick={() => onOpenConsole?.(agent.id)}
            >
              Consola
            </button>
            <button
              type="button"
              className="presence-action is-ghost"
              onClick={() => onRefreshAgent?.(agent.id)}
              title="Reevaluar disponibilidad"
            >
              ↻
            </button>
          </div>
        </li>
      );
    })}
  </ul>
);
