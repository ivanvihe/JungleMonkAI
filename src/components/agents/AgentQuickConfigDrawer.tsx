import React, { useMemo } from 'react';
import { Badge, Drawer, Space, Switch, Tag, Typography } from 'antd';
import type { BadgeProps } from 'antd';
import { RobotOutlined } from '@ant-design/icons';
import { useAgents } from '../../core/agents/AgentContext';
import { getAgentDisplayName } from '../../utils/agentDisplay';
import type { AgentDefinition } from '../../core/agents/agentRegistry';
import { ProListPanel } from '../pro';

const STATUS_COLORS: Record<string, BadgeProps['status']> = {
  Disponible: 'success',
  Inactivo: 'default',
  'Sin clave': 'warning',
  Cargando: 'processing',
  Error: 'error',
};

interface AgentQuickConfigDrawerProps {
  open: boolean;
  onClose: () => void;
}

interface AgentItem {
  id: string;
  agent: AgentDefinition;
  title: string;
  status: string;
  description?: string;
  kind: AgentDefinition['kind'];
  active: boolean;
}

export const AgentQuickConfigDrawer: React.FC<AgentQuickConfigDrawerProps> = ({ open, onClose }) => {
  const { agents, toggleAgent } = useAgents();

  const agentItems = useMemo<AgentItem[]>(
    () =>
      agents.map(agent => ({
        id: agent.id,
        agent,
        title: getAgentDisplayName(agent),
        status: agent.status,
        description: agent.objective ?? agent.description,
        kind: agent.kind,
        active: agent.active,
      })),
    [agents],
  );

  return (
    <Drawer
      title={
        <Space size={8} align="center">
          <RobotOutlined />
          <Typography.Text strong>Agentes operativos</Typography.Text>
        </Space>
      }
      placement="right"
      width={420}
      onClose={onClose}
      open={open}
      destroyOnClose={false}
      className="agent-quick-config"
    >
      <Typography.Paragraph type="secondary">
        Activa o pausa agentes rápidamente. Los cambios se aplican de forma inmediata y se sincronizan con la
        orquestación activa.
      </Typography.Paragraph>

      <ProListPanel<AgentItem>
        dataSource={agentItems}
        pagination={false}
        metas={{
          title: {
            dataIndex: 'title',
          },
          subTitle: {
            render: (_, row) => (
              <Space size={6} wrap>
                <Tag color={row.kind === 'cloud' ? 'geekblue' : 'green'}>{
                  row.kind === 'cloud' ? 'Cloud' : 'Local'
                }</Tag>
                <Badge status={STATUS_COLORS[row.status] ?? 'default'} text={row.status} />
              </Space>
            ),
          },
          description: {
            dataIndex: 'description',
            render: (_, row) =>
              row.description ? (
                <Typography.Text type="secondary">{row.description}</Typography.Text>
              ) : (
                <Typography.Text type="secondary">Sin descripción</Typography.Text>
              ),
          },
          actions: {
            render: (_, row) => [
              <Switch
                key="toggle"
                checked={row.active}
                checkedChildren="Activo"
                unCheckedChildren="Inactivo"
                onChange={() => toggleAgent(row.id)}
              />,
            ],
          },
        }}
      />
    </Drawer>
  );
};

export default AgentQuickConfigDrawer;
