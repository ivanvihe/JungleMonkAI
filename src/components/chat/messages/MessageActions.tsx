import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { AgentPresenceList } from '../../agents/AgentPresenceList';
import { useAgents } from '../../../core/agents/AgentContext';
import type { AgentPresenceEntry } from '../../../core/agents/presence';
import { useRepoWorkflow } from '../../../core/codex';

interface MessageActionsProps {
  messageId: string;
  value: string;
  onAppend?: (value: string) => void;
  onShare?: (agentId: string, canonicalCode?: string) => void;
}

export const MessageActions: React.FC<MessageActionsProps> = ({
  messageId,
  value,
  onAppend,
  onShare,
}) => {
  const containerRef = useRef<HTMLDivElement | null>(null);
  const { agents } = useAgents();
  const { queueRequest } = useRepoWorkflow();
  const [isPickerOpen, setIsPickerOpen] = useState(false);

  const handleCopy = useCallback(() => {
    const clipboard = typeof window !== 'undefined' ? window.navigator?.clipboard : undefined;
    if (clipboard?.writeText) {
      clipboard.writeText(value).catch(() => {
        // Silently ignore clipboard errors to avoid disrupting the UI
      });
    }
  }, [value]);

  const handleAppend = useCallback(() => {
    if (onAppend) {
      onAppend(value);
    }
  }, [onAppend, value]);

  const handleTogglePicker = useCallback(() => {
    if (!onShare) {
      return;
    }
    setIsPickerOpen(prev => !prev);
  }, [onShare]);

  const shareableAgents = useMemo(() => agents.filter(agent => agent.active), [agents]);

  const noopToggleAgent = useCallback((_: string) => {}, []);
  const noopUpdateRole = useCallback(
    (_agentId: string, _updates: { role?: string; objective?: string }) => {},
    [],
  );

  const presenceMap = useMemo(() => {
    const now = Date.now();
    const entries = new Map<string, AgentPresenceEntry>();
    shareableAgents.forEach(agent => {
      entries.set(agent.id, {
        status: 'online',
        lastChecked: now,
        message: 'Disponible para recibir mensajes compartidos',
      });
    });
    return entries;
  }, [shareableAgents]);

  const handleShareWithAgent = useCallback(
    (agentId: string) => {
      if (!onShare) {
        return;
      }
      onShare(agentId, value);
      setIsPickerOpen(false);
    },
    [onShare, value],
  );

  const handleSendToRepoStudio = useCallback(() => {
    if (!value.trim()) {
      return;
    }
    queueRequest({ messageId, canonicalCode: value });
  }, [messageId, queueRequest, value]);

  useEffect(() => {
    if (!isPickerOpen) {
      return;
    }

    const handleClickOutside = (event: MouseEvent) => {
      if (!containerRef.current) {
        return;
      }
      if (!containerRef.current.contains(event.target as Node)) {
        setIsPickerOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [isPickerOpen]);

  return (
    <div ref={containerRef} className="message-actions" role="group" aria-label="Acciones del bloque de código">
      <button type="button" className="message-action" onClick={handleCopy}>
        Copiar
      </button>
      {onAppend ? (
        <button type="button" className="message-action" onClick={handleAppend}>
          Añadir al compositor
        </button>
      ) : null}
      {onShare ? (
        <button type="button" className="message-action" onClick={handleTogglePicker}>
          Enviar a…
        </button>
      ) : null}
      <button type="button" className="message-action" onClick={handleSendToRepoStudio}>
        Enviar a Repo Studio
      </button>
      {onShare && isPickerOpen ? (
        <div className="message-action-popover" role="dialog" aria-label={`Compartir mensaje ${messageId}`}>
          {shareableAgents.length === 0 ? (
            <p className="message-action-empty">No hay agentes activos para compartir.</p>
          ) : (
            <AgentPresenceList
              agents={shareableAgents}
              presence={presenceMap}
              onToggleAgent={noopToggleAgent}
              onUpdateRole={noopUpdateRole}
              selectionMode
              onSelectAgent={handleShareWithAgent}
            />
          )}
        </div>
      ) : null}
    </div>
  );
};

export default MessageActions;
