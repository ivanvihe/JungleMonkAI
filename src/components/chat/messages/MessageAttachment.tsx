import React from 'react';
import { ChatAttachment, ChatTranscription } from '../../../core/messages/messageTypes';
import { AudioPlayer } from './AudioPlayer';

interface MessageAttachmentProps {
  attachment: ChatAttachment;
  transcriptions?: ChatTranscription[];
}

export const MessageAttachment: React.FC<MessageAttachmentProps> = ({ attachment, transcriptions }) => {
  if (attachment.type === 'image' && attachment.url) {
    return (
      <figure className="message-attachment image-attachment">
        <img src={attachment.url} alt={attachment.name ?? 'Imagen adjunta'} />
        {attachment.name && <figcaption>{attachment.name}</figcaption>}
      </figure>
    );
  }

  if (attachment.type === 'audio' && attachment.url) {
    const relatedTranscriptions = transcriptions?.filter(item => item.attachmentId === attachment.id);
    return (
      <div className="message-attachment audio-attachment">
        <AudioPlayer src={attachment.url} title={attachment.name} mimeType={attachment.mimeType} transcriptions={relatedTranscriptions} />
      </div>
    );
  }

  return (
    <div className="message-attachment file-attachment">
      {attachment.url ? (
        <a href={attachment.url} download={attachment.name} target="_blank" rel="noreferrer">
          {attachment.name ?? 'Archivo adjunto'}
        </a>
      ) : (
        <span>{attachment.name ?? 'Archivo adjunto'}</span>
      )}
      {attachment.mimeType && <span className="attachment-meta">{attachment.mimeType}</span>}
    </div>
  );
};

export default MessageAttachment;
