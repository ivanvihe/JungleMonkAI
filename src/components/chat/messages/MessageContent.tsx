import React from 'react';
import hljs from 'highlight.js/lib/core';
import javascript from 'highlight.js/lib/languages/javascript';
import typescript from 'highlight.js/lib/languages/typescript';
import python from 'highlight.js/lib/languages/python';
import bash from 'highlight.js/lib/languages/bash';
import json from 'highlight.js/lib/languages/json';
import xml from 'highlight.js/lib/languages/xml';
import yaml from 'highlight.js/lib/languages/yaml';
import markdown from 'highlight.js/lib/languages/markdown';
import { copyToClipboard } from '../../../utils/clipboard';
import {
  ChatContentPart,
  ChatTranscription,
} from '../../../core/messages/messageTypes';
import {
  normalizeContentParts,
  splitMarkdownContent,
} from '../../../core/messages/format';
import { AudioPlayer } from './AudioPlayer';
import { MessageActions } from './MessageActions';

hljs.registerLanguage('javascript', javascript);
hljs.registerLanguage('typescript', typescript);
hljs.registerLanguage('python', python);
hljs.registerLanguage('bash', bash);
hljs.registerLanguage('json', json);
hljs.registerLanguage('xml', xml);
hljs.registerLanguage('yaml', yaml);
hljs.registerLanguage('yml', yaml);
hljs.registerLanguage('markdown', markdown);

const escapeHtml = (code: string): string =>
  code.replace(/[&<>]/g, char => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[char] ?? char));

const highlightCode = (code: string, language?: string): string => {
  const normalized = language?.toLowerCase();
  if (normalized && hljs.getLanguage(normalized)) {
    return hljs.highlight(code, { language: normalized }).value;
  }

  try {
    return hljs.highlightAuto(code).value;
  } catch (error) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn('Fallo al resaltar código', error);
    }
    return escapeHtml(code);
  }
};

const formatLineCount = (code: string): string => {
  const lines = code.split(/\n/).length;
  return `${lines} ${lines === 1 ? 'línea' : 'líneas'}`;
};

interface MessageContentProps {
  messageId: string;
  content: string | ChatContentPart[];
  transcriptions?: ChatTranscription[];
  onAppendToComposer?: (value: string) => void;
  onShare?: (agentId: string, canonicalCode?: string) => void;
}

export const MessageContent: React.FC<MessageContentProps> = ({
  messageId,
  content,
  transcriptions,
  onAppendToComposer,
  onShare,
}) => {
  const parts = normalizeContentParts(content);

  return (
    <>
      {parts.map((part, index) => {
        if (part.type === 'text') {
          const segments = splitMarkdownContent(part.text);
          if (!segments.length) {
            return null;
          }

          return segments.map((segment, segmentIndex) => {
            if (segment.kind === 'code') {
              const highlighted = highlightCode(segment.code, segment.language);
              const lineLabel = formatLineCount(segment.code);
              const languageClass = segment.language
                ? `language-${segment.language.trim().toLowerCase()}`
                : undefined;

              return (
                <div className="message-code-block" key={`code-${index}-${segmentIndex}`}>
                  <div className="message-code-toolbar">
                    <div className="message-code-meta">
                      {segment.language ? (
                        <span className="message-code-language">{segment.language}</span>
                      ) : (
                        <span className="message-code-language">texto</span>
                      )}
                      <span className="message-code-lines">{lineLabel}</span>
                    </div>
                    <div className="message-code-actions">
                      <button
                        type="button"
                        className="message-code-copy"
                        onClick={() => {
                          void copyToClipboard(segment.code);
                        }}
                        aria-label="Copiar bloque de código"
                        title="Copiar bloque de código"
                      >
                        Copiar
                      </button>
                      <MessageActions
                        messageId={messageId}
                        value={segment.code}
                        onAppend={onAppendToComposer}
                        onShare={onShare}
                      />
                    </div>
                  </div>
                  <pre>
                    <code className={languageClass} dangerouslySetInnerHTML={{ __html: highlighted }} />
                  </pre>
                </div>
              );
            }

            return (
              <p key={`text-${index}-${segmentIndex}`} className="message-card-content">
                {segment.text}
              </p>
            );
          });
        }

        if (part.type === 'image') {
          return (
            <figure key={`image-${index}`} className="message-card-media">
              <img src={part.url} alt={part.alt ?? 'Imagen generada'} />
              {part.alt && <figcaption>{part.alt}</figcaption>}
            </figure>
          );
        }

        if (part.type === 'audio') {
          const relatedTranscriptions = transcriptions?.filter(item => !item.attachmentId);
          return (
            <div key={`audio-${index}`} className="message-card-media">
              <AudioPlayer src={part.url} title="Respuesta de audio" transcriptions={relatedTranscriptions} />
            </div>
          );
        }

        if (part.type === 'file') {
          return (
            <div key={`file-${index}`} className="message-card-media">
              <a href={part.url} target="_blank" rel="noreferrer">
                {part.name ?? 'Archivo'}
              </a>
            </div>
          );
        }

        return null;
      })}
    </>
  );
};

export default MessageContent;
