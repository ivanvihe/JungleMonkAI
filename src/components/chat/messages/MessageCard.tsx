import React from 'react';
import { ChatMessage } from '../../../core/messages/messageTypes';
import { MessageContent } from './MessageContent';
import { MessageAttachment } from './MessageAttachment';

interface MessageCardProps {
  message: ChatMessage;
  chipColor: string;
  agentDisplayName?: string;
  providerLabel?: string;
  formatTimestamp: (timestamp: string) => string;
  onAppendToComposer?: (value: string) => void;
}

export const MessageCard: React.FC<MessageCardProps> = ({
  message,
  chipColor,
  agentDisplayName,
  providerLabel,
  formatTimestamp,
  onAppendToComposer,
}) => {
  const isUser = message.author === 'user';
  const isSystem = message.author === 'system';
  const authorLabel = isUser ? 'Tú' : isSystem ? 'Control Hub' : agentDisplayName ?? 'Agente';

  return (
    <div
      className={`message-card ${isUser ? 'message-user' : ''} ${isSystem ? 'message-system' : ''}`}
      data-author={message.author}
    >
      <div className="message-card-header">
        <div className="message-card-author" style={{ borderColor: chipColor }}>
          {authorLabel}
        </div>
        <div className="message-card-meta">
          {!isUser && !isSystem && providerLabel ? (
            <span className="message-card-tag" style={{ color: chipColor }}>
              {providerLabel}
            </span>
          ) : null}
          <span className="message-card-time">{formatTimestamp(message.timestamp)}</span>
          {message.status === 'pending' && <span className="message-card-status">orquestando…</span>}
        </div>
      </div>

      <div className="message-card-body">
        <MessageContent
          content={message.content}
          transcriptions={message.transcriptions}
          onAppendToComposer={!isUser ? onAppendToComposer : undefined}
        />
        {message.attachments?.length ? (
          <div className="message-card-attachments">
            {message.attachments.map(attachment => (
              <MessageAttachment
                key={attachment.id}
                attachment={attachment}
                transcriptions={message.transcriptions}
              />
            ))}
          </div>
        ) : null}
      </div>

      {message.modalities?.length ? (
        <div className="message-card-modalities">
          {message.modalities.map(modality => (
            <span key={modality} className="modality-chip">
              {modality}
            </span>
          ))}
        </div>
      ) : null}
    </div>
  );
};

export default MessageCard;
