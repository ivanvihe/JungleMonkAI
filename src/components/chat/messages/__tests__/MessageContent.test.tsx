import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { MessageContent } from '../MessageContent';
import { AgentProvider } from '../../../../core/agents/AgentContext';
import { MessageProvider } from '../../../../core/messages/MessageContext';
import { RepoWorkflowProvider } from '../../../../core/codex';
import { PluginHostProvider } from '../../../../core/plugins/PluginHostProvider';
import { DEFAULT_GLOBAL_SETTINGS } from '../../../../utils/globalSettings';

const noop = vi.fn();

const ProviderHarness: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [settings, setSettings] = React.useState(() => ({
    ...JSON.parse(JSON.stringify(DEFAULT_GLOBAL_SETTINGS)),
  }));

  return (
    <AgentProvider apiKeys={{}}>
      <PluginHostProvider settings={settings} onSettingsChange={setSettings}>
        <MessageProvider apiKeys={{}}>
          <RepoWorkflowProvider>{children}</RepoWorkflowProvider>
        </MessageProvider>
      </PluginHostProvider>
    </AgentProvider>
  );
};

describe('MessageContent', () => {
  it('renders plain text segments', () => {
    const { container } = render(
      <ProviderHarness>
        <MessageContent messageId="message-1" content="Hola mundo" onAppendToComposer={noop} />
      </ProviderHarness>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders fenced code blocks with actions', () => {
    const codeMessage = `Respuesta:\n\n\`\`\`ts\nconst saludo: string = 'hola';\nconsole.log(saludo);\n\`\`\`\n\nFin.`;
    const { container } = render(
      <ProviderHarness>
        <MessageContent messageId="message-2" content={codeMessage} onAppendToComposer={noop} />
      </ProviderHarness>,
    );
    expect(container).toMatchSnapshot();
  });
});
