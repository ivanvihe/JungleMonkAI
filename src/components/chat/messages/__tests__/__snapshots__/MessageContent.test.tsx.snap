// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageContent > renders fenced code blocks with actions 1`] = `
<div>
  <p
    class="message-card-content"
  >
    Respuesta:


  </p>
  <div
    class="message-code-block"
  >
    <div
      class="message-code-toolbar"
    >
      <div
        class="message-code-meta"
      >
        <span
          class="message-code-language"
        >
          ts
        </span>
        <span
          class="message-code-lines"
        >
          2 líneas
        </span>
      </div>
      <div
        class="message-code-actions"
      >
        <button
          aria-label="Copiar bloque de código"
          class="message-code-copy"
          title="Copiar bloque de código"
          type="button"
        >
          Copiar
        </button>
        <div
          aria-label="Acciones adicionales del bloque de código"
          class="message-actions"
          role="group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="Abrir menú de acciones"
            class="message-action-trigger"
            type="button"
          >
            ⋯
          </button>
        </div>
      </div>
    </div>
    <pre>
      <code
        class="language-ts"
      >
        <span
          class="hljs-keyword"
        >
          const
        </span>
         
        <span
          class="hljs-attr"
        >
          saludo
        </span>
        : 
        <span
          class="hljs-built_in"
        >
          string
        </span>
         = 
        <span
          class="hljs-string"
        >
          'hola'
        </span>
        ;

        <span
          class="hljs-variable language_"
        >
          console
        </span>
        .
        <span
          class="hljs-title function_"
        >
          log
        </span>
        (saludo);
      </code>
    </pre>
  </div>
  <p
    class="message-card-content"
  >
    

Fin.
  </p>
</div>
`;

exports[`MessageContent > renders plain text segments 1`] = `
<div>
  <p
    class="message-card-content"
  >
    Hola mundo
  </p>
</div>
`;
